modules/nextflow/src/main/groovy/nextflow/script/parser/v2/ScriptCompiler.groovy

```diff
@@ -0,0 +1,295 @@
+import org.codehaus.groovy.control.Phases
+import org.codehaus.groovy.control.SourceUnit
+import org.codehaus.groovy.control.customizers.ImportCustomizer
+import org.codehaus.groovy.control.io.FileReaderSource
+import org.codehaus.groovy.control.io.ReaderSource
+import org.codehaus.groovy.control.io.StringReaderSource
+import org.codehaus.groovy.control.messages.SyntaxErrorMessage
+
+/**
+ * Compile a Nextflow script into a Groovy class.
+ *
+ * @see groovy.lang.GroovyShell::parse()
+ * @see groovy.lang.GroovyClassLoader::doParseClass()
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+@CompileStatic
+public class ScriptCompiler {
+
+    public static final String DEFAULT_CODE_BASE = "/groovy/shell"
+    public static final String MAIN_CLASS_NAME = "Main"
+
+    private final CompilerConfiguration config
+    private final GroovyClassLoader loader
+
+    private Compiler compiler
+
+    public ScriptCompiler(Session session) {
+        this(getConfig(session), session.classLoader)
+    }
+
+    public ScriptCompiler(CompilerConfiguration config, ClassLoader parent) {
+        this.config = config
+        this.loader = new GroovyClassLoader(parent, config)
+    }
+
+    private static CompilerConfiguration getConfig(Session session) {
+        final importCustomizer = new ImportCustomizer()
+        importCustomizer.addImports( java.nio.file.Path.name )
+        importCustomizer.addImports( nextflow.Channel.name )
+        importCustomizer.addImports( nextflow.util.Duration.name )
+        importCustomizer.addImports( nextflow.util.MemoryUnit.name )
+        importCustomizer.addImport( "channel", nextflow.Channel.name )
+        importCustomizer.addStaticStars( nextflow.Nextflow.name )
+
+        final config = new CompilerConfiguration()
+        config.addCompilationCustomizers( importCustomizer )
+        config.setScriptBaseClass(BaseScript.class.getName())
+        config.setPluginFactory(new ScriptParserPluginFactory())
+
+        if( session?.debug )
+            config.debug = true
+
+        if( session?.classesDir )
+            config.setTargetDirectory(session.classesDir.toFile())
+
+        return config
+    }
+
+    public CompileResult compile(String scriptText) {
+        return compile0(new GroovyCodeSource(scriptText, MAIN_CLASS_NAME, DEFAULT_CODE_BASE))
+    }
+
+    public CompileResult compile(File file) {
+        return compile0(new GroovyCodeSource(file, config.getSourceEncoding()))
+    }
+
+    public Collection<SourceUnit> getSources() {
+        if( !compiler )
+            return null
+        return compiler.getSources().values()
+    }
+
+    public List<SyntaxErrorMessage> getErrors() {
+        if( !compiler )
+            return null
+        return compiler.compilationUnit()
+            .getErrorCollector()
+            .getErrors()
+            .stream()
+            .filter(e -> e instanceof SyntaxErrorMessage)
+            .map(e -> (SyntaxErrorMessage) e)
```

Comment:
This looks just to cast the type, find a bit useless

Reply:
It doesn't compile without the type cast 😄

Reply:
```
        // collect script classes
        final classes = (collector.getLoadedClasses() as Collection<Class>)
            .findAll((Class c) -> c.getSuperclass() == BaseScript.class)
            .toList()
```

this works

Reply:
I would rather not use hard casts or Groovy `findAll`, which copies the list and converts the lambda into a closure

I'm going to leave it as it is. Stream operators are cheap, they don't make an extra copy like Groovy `collect` or `findAll` does